### PR TITLE
added gates to rotate around the X, Y and Z axes

### PR DESCRIPTION
**Pull request by craziii**
using https://www.quantum-inspire.com/kbase/rotation-operators/ as reference

### CODE REVIEW

**Comment by johanvos** [comment]
Thanks for this PR, those are useful additions indeed. It would be great if you can create a few tests for this? Similar to e.g. https://github.com/redfx-quantum/strange/blob/master/src/test/java/org/redfx/strange/test/SingleQubitGateTests.java

**Comment by craziii** [comment]
Rotation gates around the X and Y axes working as intended, However the gate to rotate around the Z axis causes an index out of bounds error in https://github.com/redfx-quantum/strange/blob/master/src/main/java/org/redfx/strange/Result.java within line 164.

This issue is nullified only be initialising the Z Gate's matrix incorrectly as [1+e^-theta/2 , 0 ; 0 , 1+e^theta/2], this error seems to derive from initialising a matrix in which some of the values are purely complex with no real component.

Upon testing with multiple values replacing the 1 within the matrix with other values the out of bounds exception only triggers for values between -1 and 1, any value outside of this range will be incorrect but not result in an index out of bounds error.

See comment on rotationZTest for information

File: src/main/java/org/redfx/strange/gate/RotationX.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
+package com.quantumtest;
```
Comment (johanvos) [inline src/main/java/org/redfx/strange/gate/RotationX.java:1]: package name should be org.redfx.strange.gate

File: src/main/java/org/redfx/strange/gate/RotationY.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
+package com.quantumtest;
```
Comment (johanvos) [inline src/main/java/org/redfx/strange/gate/RotationY.java:1]: package name should be org.redfx.strange.gate

**Comment by johanvos** [comment]
Thanks for the PR, and for catching the error in the matrix initialization. (hence, especially good that you have this test that exposes the issue).
I'll create a separate issue for that error. Once that is resolved, the tests here should pass. Meanwhile, I left a comment about wrong package names.

File: src/main/java/org/redfx/strange/gate/Rotation.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    private final Axes axis;
+
+    public Rotation(double theta, Axes axis, int idx) {
+        super(idx);
+        this.thetav = theta;
+        this.axis = axis;
+        switch(axis){
+            case XAxis: matrix =  new Complex[][]{{new Complex(Math.cos(theta/2),0), new Complex(0,-Math.sin(theta/2))}, {new Complex(0,-Math.sin(theta/2)), new Complex(Math.cos(theta/2),0)}}; break;
+            case YAxis: matrix =  new Complex[][]{{new Complex(Math.cos(theta/2),0),new Complex(-Math.sin(theta/2),0)}, {new Complex(Math.sin(theta/2),0),new Complex(Math.cos(theta/2))}}; break;
+            case ZAxis: matrix =  new Complex[][]{{new Complex(0,Math.pow(Math.E,-theta/2)),Complex.ZERO}, {Complex.ZERO,new Complex(0,Math.pow(Math.E,theta/2))}}; break;
```
[the remark this answers is not in the document]
Comment (craziii) [reply src/main/java/org/redfx/strange/gate/Rotation.java:18]: The tricky thing with the RotationZ gate is that my test will also pass for anything that does not change the outcome of the qubit, so without looking at the phase of the qubit it is functionally identical to the identity gate when measuring.

Although I am sure about the Rx and Ry gates, I do not see myself fully qualified to be certain of my implementation of the Rz gate. There is a method to create them purely using the Pauli X, Y and Z gates but I can't for the life of me find it.

[Rz Gate](https://i.imgur.com/hBjHP0a.png)
[Closest Paper on quantum gates I could find](https://threeplusone.com/pubs/on_gates.pdf)

I think after these recent changes, this is where I step out of this. I have added you as a contributer to my fork if this is the correct matrix so you can copy the line in.

[the remark this answers is not in the document]
Comment (johanvos) [reply src/main/java/org/redfx/strange/gate/Rotation.java:18]: Thanks for that contribution, really valuable.
You are correct, we don't take the phase of the qubit into account in testing. That needs to be fixed if we want to do proper testing for e.g. Rotation operators.
I realized the existing R Gate is similar to the Rz you propose here, and I am om with the concept of Rx/Ry/Rz so I will have a look at the impact of replacing R by Rz.
I'll come back to this PR after that, but I assume it is indeed just a matter of editing that matrix.

**Comment by johanvos** [comment]
Your invitation has expired. However, I'll merge this PR and create a PR immediately after that with the correct Z-matrix.
Thanks for your contribution!

**Review by johanvos — APPROVED** [review]

**Pull request merged**